File: backend/src/main/java/com/monumento/service/AuthService.java
package com.monumento.service;

import com.monumento.dto.AuthRequest;
import com.monumento.dto.AuthResponse;
import com.monumento.dto.UserDTO;
import com.monumento.model.User;
import com.monumento.repository.UserRepository;
import com.monumento.security.JwtTokenProvider;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
public class AuthService {

    @Autowired
    private UserRepository userRepository;

    @Autowired
    private PasswordEncoder passwordEncoder;

    @Autowired
    private JwtTokenProvider jwtTokenProvider;

    @Transactional
    public AuthResponse register(AuthRequest request) {
        // Check if user already exists
        if (userRepository.existsByEmail(request.getEmail())) {
            throw new RuntimeException("Email already in use");
        }

        // Create new user
        User user = User.builder()
                .email(request.getEmail())
                .passwordHash(passwordEncoder.encode(request.getPassword()))
                .subscriptionTier("free")
                .build();

        user = userRepository.save(user);

        // Generate tokens
        String accessToken = jwtTokenProvider.generateAccessToken(user.getId(), user.getEmail());
        String refreshToken = jwtTokenProvider.generateRefreshToken(user.getId());

        // Build response
        UserDTO userDTO = UserDTO.builder()
                .id(user.getId())
                .email(user.getEmail())
                .fullName(user.getFullName())
                .avatarUrl(user.getAvatarUrl())
                .subscriptionTier(user.getSubscriptionTier())
                .createdAt(user.getCreatedAt())
                .build();

        return AuthResponse.builder()
                .accessToken(accessToken)
                .refreshToken(refreshToken)
                .tokenType("Bearer")
                .expiresIn(86400L) // 24 hours in seconds
                .user(userDTO)
                .build();
    }

    @Transactional(readOnly = true)
    public AuthResponse login(AuthRequest request) {
        // Find user by email
        User user = userRepository.findActiveByEmail(request.getEmail())
                .orElseThrow(() -> new RuntimeException("Invalid email or password"));

        // Verify password
        if (!passwordEncoder.matches(request.getPassword(), user.getPasswordHash())) {
            throw new RuntimeException("Invalid email or password");
        }

        // Generate tokens
        String accessToken = jwtTokenProvider.generateAccessToken(user.getId(), user.getEmail());
        String refreshToken = jwtTokenProvider.generateRefreshToken(user.getId());

        // Build response
        UserDTO userDTO = UserDTO.builder()
                .id(user.getId())
                .email(user.getEmail())
                .fullName(user.getFullName())
                .avatarUrl(user.getAvatarUrl())
                .subscriptionTier(user.getSubscriptionTier())
                .createdAt(user.getCreatedAt())
                .build();

        return AuthResponse.builder()
                .accessToken(accessToken)
                .refreshToken(refreshToken)
                .tokenType("Bearer")
                .expiresIn(86400L)
                .user(userDTO)
                .build();
    }

    @Transactional(readOnly = true)
    public AuthResponse refreshToken(String refreshToken) {
        // Validate refresh token
        if (!jwtTokenProvider.validateToken(refreshToken)) {
            throw new RuntimeException("Invalid refresh token");
        }

        // Get user ID from token
        var userId = jwtTokenProvider.getUserIdFromToken(refreshToken);

        // Find user
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new RuntimeException("User not found"));

        // Generate new access token
        String newAccessToken = jwtTokenProvider.generateAccessToken(user.getId(), user.getEmail());
        String newRefreshToken = jwtTokenProvider.generateRefreshToken(user.getId());

        // Build response
        UserDTO userDTO = UserDTO.builder()
                .id(user.getId())
                .email(user.getEmail())
                .fullName(user.getFullName())
                .avatarUrl(user.getAvatarUrl())
                .subscriptionTier(user.getSubscriptionTier())
                .createdAt(user.getCreatedAt())
                .build();

        return AuthResponse.builder()
                .accessToken(newAccessToken)
                .refreshToken(newRefreshToken)
                .tokenType("Bearer")
                .expiresIn(86400L)
                .user(userDTO)
                .build();
    }
}

File: frontend-web/src/lib/api/auth.service.ts
/**
 * Authentication Service
 * Handles user registration, login, and token management
 */

import { apiClient } from './client';

export interface AuthRequest {
  email: string;
  password: string;
}

export interface User {
  id: string;
  email: string;
  fullName?: string;
  avatarUrl?: string;
  subscriptionTier: string;
  createdAt: string;
}

export interface AuthResponse {
  accessToken: string;
  refreshToken: string;
  tokenType: string;
  expiresIn: number;
  user: User;
}

/**
 * Register a new user
 */
export async function register(email: string, password: string): Promise<AuthResponse> {
  const response = await apiClient.post<AuthResponse>('/auth/register', {
    email,
    password,
  });

  // Store tokens in localStorage
  if (response) {
    localStorage.setItem('auth', JSON.stringify(response));
  }

  return response;
}

/**
 * Login with email and password
 */
export async function login(email: string, password: string): Promise<AuthResponse> {
  const response = await apiClient.post<AuthResponse>('/auth/login', {
    email,
    password,
  });

  // Store tokens in localStorage
  if (response) {
    localStorage.setItem('auth', JSON.stringify(response));
  }

  return response;
}

/**
 * Refresh access token using refresh token
 */
export async function refreshToken(): Promise<AuthResponse> {
  const authData = localStorage.getItem('auth');
  if (!authData) {
    throw new Error('No auth data found');
  }

  const { refreshToken } = JSON.parse(authData);
  if (!refreshToken) {
    throw new Error('No refresh token found');
  }

  const response = await apiClient.post<AuthResponse>('/auth/refresh', {
    refreshToken,
  });

  // Update stored tokens
  if (response) {
    localStorage.setItem('auth', JSON.stringify(response));
  }

  return response;
}

/**
 * Logout user (client-side only - JWT is stateless)
 */
export function logout(): void {
  localStorage.removeItem('auth');
}

/**
 * Get current user from stored auth data
 */
export function getCurrentUser(): User | null {
  const authData = localStorage.getItem('auth');
  if (!authData) return null;

  try {
    const parsed: AuthResponse = JSON.parse(authData);
    return parsed.user || null;
  } catch {
    return null;
  }
}

/**
 * Check if user is authenticated
 */
export function isAuthenticated(): boolean {
  const authData = localStorage.getItem('auth');
  if (!authData) return false;

  try {
    const parsed: AuthResponse = JSON.parse(authData);
    return !!parsed.accessToken;
  } catch {
    return false;
  }
}

export const authService = {
  register,
  login,
  refreshToken,
  logout,
  getCurrentUser,
  isAuthenticated,
};

export default authService;
